metaflow/task.py

```diff
@@ -439,7 +410,43 @@ def run_step(self,
                     # initialize parameters (if they exist)
                     # We take Parameter values from the first input,
                     # which is always safe since parameters are read-only
-                    current._update_env({'parameter_names': self._init_parameters(inputs[0])})
+                    current._update_env(
+                        {'parameter_names': self._init_parameters(inputs[0])})
+
+            # Note: All internal flow attributes (ie: non-user artifacts)
+            # should either be set prior to running the user code or listed in
+            # FlowSpec._EPHEMERAL to allow for proper merging/importing of
+            # user artifacts in the user's step code.
+            decorators = step_func.decorators
+            for deco in decorators:
+
+                deco.task_pre_step(step_name,
+                                   output,
+                                   self.metadata,
+                                   run_id,
+                                   task_id,
+                                   self.flow,
+                                   self.flow._graph,
+                                   retry_count,
+                                   max_user_code_retries,
+                                   self.ubf_context,
+                                   inputs)
+
+                # decorators can actually decorate the step function,
+                # or they can replace it altogether. This functionality
+                # is used e.g. by catch_decorator which switches to a
+                # fallback code if the user code has failed too many
+                # times.
+                step_func = deco.task_decorate(step_func,
+                                               self.flow,
+                                               self.flow._graph,
+                                               retry_count,
+                                               max_user_code_retries,
+                                               self.ubf_context)
+
+            if join_type:
+                self._exec_step_function(step_func, input_obj)
```

Comment:
Any reason to move the code around? `input_obj` only exists `if join_type` so it makes sense to invoke `_exec_step_function` where `input_obj` is defined.

Reply:
I will check. This is how we do it internally so kept that to avoid breaking things.

Reply:
I checked. The difference is that with this new approach, the datastore is properly set and you can access the attributes in the flow properly which is needed in one of the internal plugins. In fairness, we also will need to move this around for the new datastore since we pass the datastore down to task_pre_step.

Reply:
pre_step and decorate don't rely on datastore at the moment. I am not sure how this change is functionally different from the tip of the master.